lightning/src/ln/onion_utils.rs

```diff
@@ -592,6 +592,145 @@ pub(super) fn process_onion_failure<T: secp256k1::Signing, L: Deref>(secp_ctx: &
+		else if failure_code == 18 { debug_assert_eq!(data.len(), 4) }
+		else if failure_code == 19 { debug_assert_eq!(data.len(), 8) }
+		else if failure_code == 20 | UPDATE {
+			debug_assert_eq!(data.len() - 2 - 2, u16::from_be_bytes(data[2..4].try_into().unwrap()) as usize) }
+		else if failure_code == 21 { debug_assert!(data.is_empty()) }
+		else if failure_code == 22 | PERM { debug_assert!(data.len() <= 11) }
+		else if failure_code == 23 { debug_assert!(data.is_empty()) }
+		else if failure_code & BADONION != 0 {
+			// We set some bogus BADONION failure codes in test, so ignore unknown ones.
+		}
+		else { debug_assert!(false, "Unknown failure code: {}", failure_code) }
+
+		Self(HTLCFailReasonRepr::Reason { failure_code, data })
+	}
+
+	pub(super) fn from_failure_code(failure_code: u16) -> Self {
+		Self::reason(failure_code, Vec::new())
+	}
+
+	pub(super) fn from_msg(msg: &msgs::UpdateFailHTLC) -> Self {
+		Self(HTLCFailReasonRepr::LightningError { err: msg.reason.clone() })
+	}
+
+	pub(super) fn get_encrypted_failure_packet(&self, incoming_packet_shared_secret: &[u8; 32], phantom_shared_secret: &Option<[u8; 32]>)
+	-> msgs::OnionErrorPacket {
+		match self.0 {
+			HTLCFailReasonRepr::Reason { ref failure_code, ref data } => {
+				if let Some(phantom_ss) = phantom_shared_secret {
+					let phantom_packet = build_failure_packet(phantom_ss, *failure_code, &data[..]).encode();
+					let encrypted_phantom_packet = encrypt_failure_packet(phantom_ss, &phantom_packet);
+					encrypt_failure_packet(incoming_packet_shared_secret, &encrypted_phantom_packet.data[..])
+				} else {
+					let packet = build_failure_packet(incoming_packet_shared_secret, *failure_code, &data[..]).encode();
+					encrypt_failure_packet(incoming_packet_shared_secret, &packet)
+				}
+			},
+			HTLCFailReasonRepr::LightningError { ref err } => {
+				encrypt_failure_packet(incoming_packet_shared_secret, &err.data)
+			}
+		}
+	}
+
+	pub(super) fn decode_onion_failure<T: secp256k1::Signing, L: Deref>(
+		&self, secp_ctx: &Secp256k1<T>, logger: &L, htlc_source: &HTLCSource
+	) -> (Option<crate::routing::gossip::NetworkUpdate>, Option<u64>, bool, Option<u16>, Option<Vec<u8>>)
+	where L::Target: Logger {
+		match self.0 {
+			HTLCFailReasonRepr::LightningError { ref err } => {
+				process_onion_failure(secp_ctx, logger, &htlc_source, err.data.clone())
+			},
+			HTLCFailReasonRepr::Reason { ref failure_code, ref data, .. } => {
+				// we get a fail_malformed_htlc from the first hop
+				// TODO: We'd like to generate a NetworkUpdate for temporary
+				// failures here, but that would be insufficient as find_route
+				// generally ignores its view of our own channels as we provide them via
+				// ChannelDetails.
+				if let &HTLCSource::OutboundRoute { ref path, .. } = htlc_source {
+					(None, Some(path.first().unwrap().short_channel_id), true, Some(*failure_code), Some(data.clone()))
+				} else { unreachable!(); }
```

Comment:
It'd remove some `unreachable`s to make `HTLCSource::OutboundRoute` have an inner struct, but I think that'd be a future consideration